쥬스메이커 [STEP3] Vetto, 레옹아범

fatherLeon opened the pull request:
@llingo
안녕하세요 링고, 지난번에 많이 도와주셔서 이번에 조금 더 쉽게 PR보낼수 있었던것 같습니다
좋은 리뷰 부탁드립니다🙏👍

## 고민한점
### 데이터 공유 방식
* 많은 공유 방식이 있었고 그 중 처음에는 singleton pattern을 이용하는 식으로 코드를 작성했습니다.
* step2를 진행하고 있을 때 리뷰어에게 어떤 방식이 좋은 지 추천을 받았는데 delegate pattern도 좋다고 리뷰를 받았습니다.
* 그 답변을 받은 후 팀원과 상의한 끝에 singleton pattern도 좋지만 delegate pattern도 한번 써보는 것이 좋다고 생각하여 delegate를 선택하였고 delegate를 기반으로 코드를 작성했습니다.

### 부분 업데이트
* FruitStockViewController에서 stepper를 사용시 fruit label 전체를 업데이트 해주는 효율이 좋지 않은 코드를 작성했습니다.
* 현재 프로젝트는 많은 양을 다루지 않기 때문에 실행하는 것에 별 문제가 없지만 데이터가 많아지게 된다면 데이터를 업데이트 하는데 시간이 많이 걸릴 것이라 생각합니다.
* 따라서 부분만 업데이트를 할 수 있게 코드를 작성하였고 그로 인해 decideStepper의 로직을 바꾸고 전체적인 refactoring을 하였습니다.

수정 전
```swift
private func updateLabel(juice: [Fruit: Int]) {
    guard let strawberry = juice[.strawberry],
        let banana = juice[.banana],
        let kiwi = juice[.kiwi],
        let pineapple = juice[.pineapple],
        let mango = juice[.mango] else { return }

    strawberryLabel.text = "\(strawberry)"
    bananaLabel.text = "\(banana)"
    kiwiLabel.text = "\(kiwi)"
    pineappleLabel.text = "\(pineapple)"
    mangoLabel.text = "\(mango)"
}
```
 
수정 후
```swift
private func updateLabel(fruit: Fruit, label: UILabel) {
    guard let changedStock = getFruits[fruit] else { return }

    label.text = "\(changedStock)"
}
```

## 조언을 얻고 싶은 부분 && 해결하지 못한 점

### Delegate weak관련 부분
* Delegate를 사용할 때 해당 프로토콜을 채택한 프로퍼티에 weak를 붙여주는 것으로 알고 있습니다.
* 그 이유는 저희 프로젝트를 통해 예를들면 delegate프로퍼티가 다른 뷰 컨트롤러를 참조하고 있어 자동적으로 FruitStockViewController가 deinit이 되지 않기 때문에 weak를 붙여서 뷰가 사라지면 deinit되는 방향으로 알고 있습니다.
* 실제로 Delegate로 양방향 데이터 전달을 했을 경우 뷰가 사라졌음에도 불구하고 deinit이 되지 않는 경우가 있어서 weak를 사용하여 deinit을 시켜주는 것으로 이해했습니다.
* 그런데 현재 delegate가 단방향 데이터전달일 경우에 weak를 써주지 않았음에도 불구하고 FruitStockViewController의 뷰가 사라지면 deinit이 되는데 이부분에 대한 이해가 맞는지 궁금합니다.

![](https://docs.swift.org/swift-book/_images/referenceCycle02_2x.png)
* 양방향 통신의 경우 위와 같은 상황이라 weak을 해줘야 한다면

![스크린샷 2023-01-13 오후 6 04 46](https://user-images.githubusercontent.com/55453151/212281507-de8fd998-49e2-4cde-8a14-6747e435a873.png)
* 위 사진과 저희 프로젝트에서 구성한 내용처럼 한쪽만 강한 참조가 걸려있다면 자동으로 해당 참조값이 nil로 변하여 deinit이 된다고 이해하면 되는지 궁금합니다

fatherLeon committed:
feat: 함수 메소드 프로퍼티 생성 및 구현

updateStepper()메소드 구현 및 5개의 stepper프로퍼티 생성

feat

Commit:
feat: stepper 구현

clickStepper 구현 및 viewWillAppear 구현

feat

fatherLeon committed:
fix: stepper동작 이상 버그 수정

clickStepper내 변수 수정

fix

Commit:
feat: NavigationBar 수정

NavigationBar Title 재고 추가로 변경 및 RightBarButton 닫기 추가

feat

fatherLeon committed:
Add 메소드 구현

touchUpDismissButton 구현

Add

fatherLeon committed:
feat: 네비게이션 바 뒤로가기 버튼 삭제

fatherLeon committed:
style: 함수 위치 수정

Commit:
refactoring: 은닉화, 접근 제어 수정

FruitStockViewController 밑 ViewController final키워드 추가, IBOutlet private로 은닉화, class 내부 메서드 private로 은닉화

refactoring

fatherLeon committed:
refactor: STEP2 리뷰 수정

viewcontroller이름 변경 및 메소드 이름 변경

refactor

Commit:
refactor: decideOrderJuice함수 리팩토링

fatherLeon committed:
feat: 메소드 생성 및 static변수 생성

pushFruitStockViewController 메소드 생성 및 FruitStockViewController의 identifier static상수 생성

feat

Commit:
feat: autolayout 설정

FruitStockViewController 12pro 이외에도 화면이 짤리지 않도록 설정

feat

fatherLeon committed:
refactor: 함수명 변경

FruitStockViewController내의 IBAction함수명 변경

refactor

fatherLeon committed:
chore

Commit:
refactor: 데이터 공유 방식 변경에 따른 수정

singleton에서 delegate로 데이터 공유 방식 변경에 따른 리팩토링

refactor

fatherLeon committed:
feat: 데이터전달(delegate) 구현

StockViewController -> JuiceMakerViewController delegate로 데이터 전달 구현

feat

Commit:
refactor: 데이터 공유방식에 따른 리택토링

데이터 전달 객체를 Dictionary컬렉션에서 FruitStore 타입으로 리팩토링

refactor

fatherLeon committed:
docs: README 수정

fatherLeon committed:
docs: README 수정

fatherLeon committed:
refactor, feat: 메소드 추가 및 전체적인 네이밍 변경

updateLabel, updateStepper과 같은 메소드 이름 변경
updateLabel을 새로 생성하여 하나의 과일에 해당하는 하나의 라벨만 변경 되게끔 하는 메소드 구현

refactor feat

Commit:
refactor: 리팩토링을 위한 함수명 변경, getFruits 연산프로퍼티 생성

fatherLeon committed:
Merge branch 'ic_8_leonfather' into step3

seu0313 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -1,7 +1,7 @@
 //  Created by Vetto, 레옹아범 on 2023.1.2
 
-enum Juice {
+enum Juice: CaseIterable {
     case strawberryJuice
     case bananaJuice
     case kiwiJuice
```
CaseIterable을 사용해주셨군요 👍

seu0313 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -74,33 +77,32 @@ final class JuiceMakerViewController: UIViewController {
-        default:
-            return nil
-        }
+        return Juice.allCases.first { juice == "\($0.name) 주문" }
```
CaseIterable을 잘 활용해주셨네요 좋습니다 👍

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -14,14 +14,38 @@ final class FruitStockViewController: UIViewController {
+    var getFruits: [Fruit: Int] {
+        return fruitStore.getFruits()
+    }
+    var delegate: Delegate?
```
[P1] 단방향 데이터 전달이라도 혹시 모를 강한 참조를 방지하기 위해 Delegate 프로토콜에 AnyObject를 채택한 후 weak을 사용하여 약한 참조를 주는 것이 좋을 것 같습니다

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceProtocol.swift:
```diff
@@ -0,0 +1,6 @@
+
+protocol Delegate {
+    func updateStock(changeStock: FruitStore)
+}
```
[P1] Delegate의 네이밍을 자세하면 변경하면 좋을 것 같아요. 만약, Delegate 프로토콜이 여러 개일때 어떤 클래스의 Deleagate 프로토콜인지 알기 힘들 것 같습니다.

fatherLeon replied:
* 오늘 활동학습 시간에 Protocol과 Delegate에 대해 배우면서 프로토콜 네이밍은 ~~able을 자주 붙인다고 말씀하셔서
```swift
protocol UpdatableFruitStock: AnyObject {
}
```
UpdatableFruitStock으로 네이밍을 변경했습니다. 8c5553d1fa026bf08b4c9ab291be1f0355341cb3

seu0313 replied:
[P3] 디스코드에서 말씀드린 것처럼 Delegate 패턴을 적용한 프로토콜의 경우
일반적으로 ~Delegate, ~Listener, ~Handler 와 같은 네이밍을 하는 것 같습니다 :)

seu0313 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,7 @@
+    init(amount: Int = 0) {
         self.fruits[.strawberry] = amount
         self.fruits[.banana] = amount
         self.fruits[.kiwi] = amount
```
[P2] `private(set)` 접근제어자를 사용하면 외부에선 읽기전용이며 내부에서는 변경이 가능한 변수로 만들 수 있습니다.
getFruits 메서드를 통해 재고를 가져오는 것보다 `private(set)` 을 사용해보시면 좋을 것 같아요

생성자를 통해 딕셔너리에 과일과 재고를 추가해주고 있기 때문에 `[:]`로 초기화해도 될 것 같습니다

```swift
final class FruitStore {
    private(set) var fruits: [Fruit: Int] = [:]

    init(amount: Int = 0) {
      ...
    }
}
```

gzzjk159 replied:
리뷰어의 조언대로 `private(set)` 접근제어자를 활용하였습니다. 그로 인해 `getfruits()`를 없애고 `.fruits`로 접근하여 더 가독성 높은 코드로 리팩토링 하였습니다. ebf0455cb3b4572dab666b21d5ba58596f1172e8

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -13,8 +13,12 @@ struct JuiceMaker {
+    }
+    
+    mutating func setFruitStore(_ fruits: FruitStore) {
+        fruitStore = fruits
```
FruitStore 자체를 교체해주는 건가요? 🤔

gzzjk159 replied:
리팩토링 하기 전에는 FruitStore를 주고 받는 식으로 작성 하였으나 현재는 문제점을 파악하고 `[Fruit: Int]`를 주고 받는 코드로 리팩토링 하였습니다. b772b2249cda3ee6c9a8770c8c6f6c444317ecfa

seu0313 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -74,33 +77,32 @@ final class JuiceMakerViewController: UIViewController {
+extension JuiceMakerViewController: Delegate {
+    func updateStock(changeStock: FruitStore) {
+        self.juiceMaker.setFruitStore(changeStock)
+    }
```
[P1] FruitStore 자체를 교체해주는 것보다 `[Fruit: Int]`의 재고 상태 정보를 사용해서 업데이트 해주는 것이 좋을 것 같아요. 클래스의 특징에 대해 생각하셔서 한번 적용해보시면 좋을 것 같습니다.

gzzjk159 replied:
클래스의 특징으로는 주소를 가르키는 변수(상수)는 스택영역에 있고 실제 데이터들은 힙 영역에 있는 것입니다. 저희 코드에선 힙 영역에 FruitsStore가 두번 JuiceMakerViewController에서도 한번 할당하고 `FruitStockViewController`에서도 할당을 하고 재참조를 하기 때문에 메모리 낭비라고 생각합니다. 리뷰어의 조언대로 `FruitStore`클래스를 넘겨주는 것 보다 `[Fruit: Int]`딕셔너리를 넘겨주는 식으로 바꿨습니다. b772b2249cda3ee6c9a8770c8c6f6c444317ecfa

seu0313 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -74,33 +77,32 @@ final class JuiceMakerViewController: UIViewController {
+        guard let fruitStockViewController = self.storyboard?
+            .instantiateViewController(withIdentifier: FruitStockViewController.identifier) as? FruitStockViewController else { return }
+        
+        fruitStockViewController.setFruits(juiceMaker.getFruitStore())
```
[P1] FruitStore 자체를 넘겨주는 것보다 현재 과일 재고 정보를 주고 받는 것이 좋지 않을까 생각됩니다. 
클래스의 특징에 대해 생각하셔서 한번 적용해보시면 좋을 것 같습니다.

fatherLeon replied:
```swift
func setFruits(_ fruits: [Fruit: Int]) {
}
```
* 우선 딕셔너리로 받아서 처리하게 끔 구현하였습니다.
* FruitStore자체를 넘기는것보다 딕셔너리를 만들어서 사용하라고 하신 이유를 생각해봤습니다🤔🤔
1. 딕셔너리는 밸류타입이라 메모리 구조에서 스택에 할당되며 스택은 힙에비해 속도가 빠르기 때문에 딕셔너리를 사용하라고 하신것 같습니다
2. 클래스를 넘겨줄 경우 FruitStore쪽에 다시 관련 클래스가 힙에 할당되어 메모리 낭비이기 때문에 딕셔너리를 넘겨주라고 말씀하신것 같습니다

seu0313 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -16,13 +19,17 @@ final class JuiceMakerViewController: UIViewController {
         super.viewWillAppear(animated)
         
-        updateLabel(juice: juiceMaker.getFruitsInStore())
+        setDefaultLabel(juice: getFruits)
```
[P3] 해당 부분은 FruitStockViewController -> JuiceMakerViewController로 넘어올 때 Label들을 세팅하기 위해 구현하신 것 같네요. 실제로 업데이트를 delegate 프로토콜 메서드인 updateStock에서 갱신해주므로 그 쪽에서 Label들을 세팅하면 좋지 않을까 생각되네요.

fatherLeon replied:
* viewWillAppear()은 제거하고
```swift
func updateStock(changeStock: [Fruit: Int]) {
    self.juiceMaker.setFruitStore(changeStock)
    setLabel(changeStock)
}
```
* 프로토콜 메소드 updateStock에서 라벨을 업데이트 할 수 있게 변경하였습니다.

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -14,14 +14,38 @@ final class FruitStockViewController: UIViewController {
+    @IBOutlet weak private var kiwiStepper: UIStepper!
+    @IBOutlet weak private var mangoStepper: UIStepper!
+    
+    var fruitStore: FruitStore = FruitStore()
```
[P3] JuiceMakerViewController의 JuiceMaker 인스턴스 내부의 FruitStore와 별개의 인스턴스로 구현하신 것 같습니다.
두 인스턴스의 재고 정보를 별도의 메서드를 통해 일치시켜주고 있기 때문에 요구사항에는 충족하지만 FruitStockViewController에 FruitStore 인스턴스가 없어도 되지 않을까 생각됩니다.

gzzjk159 replied:
리뷰어 조언대로 FruitStore 클래스를 두 번 할당하는 것이 메모리 낭비라 생각하여 해당 코드를 삭제하고 리팩토링 하였습니다. b772b2249cda3ee6c9a8770c8c6f6c444317ecfa

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -14,14 +14,38 @@ final class FruitStockViewController: UIViewController {
-   private func updateLabel(juice: [Fruit: Int]) {
+    func setFruits(_ fruits: FruitStore) {
+        self.fruitStore = fruits
+    }
```
[P1] FruitStore 자체를 교체하는 것보다 `[Fruit: Int]`의 재고 정보를 주고 받는 것이 좋을 것 같네요

gzzjk159 replied:
리뷰어의 조언대로 FruitStore를 주고 받는 것보다 `[Fruit: Int]`의 재고 정보를 주고 받는 것이 더 좋다고 생각하여 팀원과 상의하여 재고 정보만 주고 받는 코드로 리팩토링 하였습니다. b772b2249cda3ee6c9a8770c8c6f6c444317ecfa

수정 전
```swift
func setFruits(_ fruits: FruitStore) {
        self.fruitStore = fruits
```
수정 후
```swift
func setFruits(_ fruits: [Fruit: Int]) {
        self.fruits = fruits
}
```

seu0313 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -34,4 +58,61 @@ final class FruitStockViewController: UIViewController {
+            fruitStore.decreaseFruit(fruit)
+        }
+        
+        updateLabel(fruit: fruit, label: label)
```
FruitStore가 아닌 재고 정보를 주고 받도록 변경한다면 위 메서드의 로직도 간단해질 것 같네요 :)

fatherLeon replied:
```swift
@IBAction private func didTapFruitStepper(_ sender: UIStepper) {
    guard let (fruit, label) = decideStepper(sender) else { return }
        
    fruits[fruit] = Int(sender.value)
        
    updateLabel(label: label, stepper: sender)
}
```
* 기존 fruitStore의 값을 변경해주고 fruitStore의 값을 라벨에 입력하는 형식에서
* stepper의 값을 통해 label을 변경해주고 딕셔너리 값을 변경해주는 식으로 구성하였습니다.

seu0313 commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -1,11 +1,14 @@
-//  JuiceMaker - JuiceMakerViewController.swift
+//  JuiceMaker - ViewController.swift
```
주석의 ViewController 이름과 파일명이 다른 것 같네요

gzzjk159 replied:
전에 푸쉬할 때 바꿨는데 머지할 때 충돌이 일어나서 수정할 때 다시 돌아온거 같습니다. 현재는 문제점을 인지 하고 해당 파일명과 일치시켰습니다. ebf0455cb3b4572dab666b21d5ba58596f1172e8

seu0313 reviewed (changes requested):
안녕하세요 vetto, 레옹아범
쥬스메이커 프로젝트 STEP 3 수고하셨습니다 :)

---

> 안내사항은 참고용으로 남겨두었습니다.

1. **PR을 보내기 전에는 3가지를 확인해주세요.**
    - `캠핑사이트 - Course - Pull Request 보낼 때` 글을 꼭 확인해주세요.
    - `yagom-academy:main`로 보내고 있는 것은 아닌 지 확인해주세요.
    - 이전 STEP이 Merge된 후 다음 STEP의 PR을 보내주세요.
2. **PR 수정 코멘트에는 P1 - P3의 중요도를 남겨두었습니다. 중요하지 않는 경우 확인했다는 이모지만 남기셔도 됩니다.**
    - [P1] 무조건 수정이 필요합니다.
    - [P2] 다음 스탭에서 수정해도 됩니다.
    - [P3] 꼭 반영하지 않고 참고만 해도 됩니다.
3. **PR 처럼 수정 후에도 DM를 남겨주시면 빠르게 확인하겠습니다.**

---

**Q1-1. Delegate weak 관련 부분**

> 현재 delegate가 단방향 데이터전달일 경우에 weak를 써주지 않았음에도 불구하고 FruitStockViewController의 뷰가 사라지면 deinit이 되는데 이부분에 대한 이해가 맞는지 궁금합니다.

A1-1. 네 delegate를 강한 참조로 구현하셨고 현재 상황을 그려보면 아래와 같은 상황인 것 같아요.

1. 앱이 시작하면서 UINavigationController의 root로 JuiceMakerVC 참조
2. FruitStockVC의 delegate를 JuiceMakerVC로 설정 및 UINavigationController로 push
3. UINavigationController로 pop
4. FruitStockVC deinit (Strong Reference Count = 0)

![](https://i.imgur.com/5ypgFxy.png)

지금 구현하신 코드에서는 강한 참조의 문제가 발생하지 않았지만 추가 기능을 구현하는 과정에서
언제든지 강한 참조가 발생할 수 있는 가능성이 존재하기 때문에 delegate를 weak으로 설정하는 것이 좋을 것 같습니다 :)

**Q1-2. Delegate weak 관련 부분**

> 위 사진과 저희 프로젝트에서 구성한 내용처럼 한쪽만 강한 참조가 걸려있다면 자동으로 해당 참조값이 nil로 변하여 deinit이 된다고 이해하면 되는지 궁금합니다

A1-2. nil로 변하여 deinit이 된 것이 아닙니다. Q1-1의 이미지를 참고하시면 좋을 것 같습니다.

fatherLeon committed:
refactor: 프로토콜 및 FruitStore 수정

프로토콜 네이밍 및 프로토콜 타입 변수 weak 변수로 바꿈
FruitStore fruits변수 초기 설정변경

refactor

Commit:
refactor: 접근제어자 변경 및 그에 따른 리팩토링

FruitStore 접근 제어자 private(set)으로 변경, getFruits삭제 및 코드 수정

refactor

fatherLeon committed:
refactor: delegate메소드 재구현

delegate 메소드인 useFruit인자 재구성 및 JuiceMaker의 setFruitsStore메소드 변경

refactor

Commit:
refactor: label 업데이트 방법 수정

JuiceMakerViewController의 업데이트 방법 전체 업데이트에서 부분 업데이트로 변경

refactoer

fatherLeon committed:
refactor: FruitStockViewController 리팩토링

기존 FruitStore타입을 가진 프로퍼티를 제거 -> fruits라는 딕셔너리만 생성

refactor

Commit:
refactor: 메서드 로직 수정, 메서드 명 수정

fatherLeon committed:
refactor: 접근제어자 설정 및 메소드명 변경

JuiceMakerViewController -> 각 프로퍼티들 private으로 접근제어자 설정
FruitStockViewController -> viewWillAppear 삭제 및 setLabelAndStepperValue로 setLabel과 setStepperValue 메소드를 합침

refactor

Commit:
refactor: required init 사용 리팩토링

fatherLeon committed:
fix: JuiceMakerViewController 프로퍼티 변경에 따른 에러 수정

Commit:
refactor: 연산 프로퍼티 수정

fatherLeon commented:
@llingo 어제 말씀하신 부분을 해결하고 또 궁금증이 있어서 코멘트 남겼습니다!!!
## 해결한 점
## 1️⃣사용하지 않는 메서드 제거

### FruitStore
```swift
func increaseFruit(_ fruit: Fruit, by amount: Int = 1) {
    guard let storedFruit = self.fruits[fruit] else {
        return
    }
    self.fruits[fruit] = storedFruit + amount
}
```
* 위에 코드는 저희가 구현은 했지만 나중에 코드 리팩토링을 하다보니 사용하지 않는 것을 깨달았습니다. 사용하지 않는 코드는 삭제하는 것이 좋다고 말씀해주신 리뷰어의 조언에 따라 삭제를 하였습니다.

### JuiceMaker
```swift
func getFruitStore() -> FruitStore {
     return self.fruitStore
}

mutating func setFruitStore(_ fruits: [Fruit: Int]) {
    fruitStore.setFruits(fruits)
}
```
* JuiceMakerViewController내에서 FruitStore로 접근이 가능해져서 JuiceMaker내의 메소드는 쥬스를 만드는 과정이나 확인 과정만을 남겨둘 수 있게되어 JuiceMaker내에서 FruitStore를 가져오거나 할당해주는 메소드는 전부 삭제처리 하였습니다.

## 2️⃣ViewController의 init
### JuiceMakerViewController
```swift
required init?(coder: NSCoder) {
    fruitStore = FruitStore(amount: 10)
    juiceMaker = JuiceMaker(fruitStore: fruitStore)
    super.init(coder: coder)
}
```
* 말씀해주신대로 각각의 인스턴스를 생성하여 init내에서 처리할 수 있도록 바꿨습니다

## 3️⃣변경한 코드 주석 삭제
* 리팩토링후 바뀐 코드와 바뀌기 전 코드 둘 중 어느 코드가 좋은지 확신을 갖지 못하여서 코드를 주석으로 남겨놨는데 리뷰어 조언대로 바뀐 코드로 수정하기로 하여 주석처리한 부분을 삭제 하였습니다.
```swift
//    private func setLabel() {
//        guard let strawberryStock = fruits[.strawberry],
//              let bananaStock = fruits[.banana],
//              let kiwiStock = fruits[.kiwi],
//              let pineappleStock = fruits[.pineapple],
//              let mangoStock = fruits[.mango] else { return }
//
//        strawberryLabel.text = "\(strawberryStock)"
//        bananaLabel.text = "\(bananaStock)"
//        kiwiLabel.text = "\(kiwiStock)"
//        pineappleLabel.text = "\(pineappleStock)"
//        mangoLabel.text = "\(mangoStock)"
//    }
//
//    private func setStepperValue() {
//        guard let strawberryStock = fruits[.strawberry],
//              let bananaStock = fruits[.banana],
//              let pineappleStock = fruits[.pineapple],
//              let kiwiStock = fruits[.kiwi],
//              let mangoStock = fruits[.mango] else { return }
//
//        strawberryStepper.value = Double(strawberryStock)
//        bananaStepper.value = Double(bananaStock)
//        pineappleStepper.value = Double(pineappleStock)
//        kiwiStepper.value = Double(kiwiStock)
//        mangoStepper.value = Double(mangoStock)
//    }
```

## 조언을 얻고 싶은 부분
### 1️⃣JuiceMakerViewController

**수정 전**
```swift
 private var strawberryAmount: Int {
    guard let amount = getFruits[.strawberry] else { return 0 }
    return amount
}
private var bananaAmount: Int {
    guard let amount = getFruits[.banana] else { return 0 }
    return amount
}
private var pineappleAmount: Int {
    guard let amount = getFruits[.pineapple] else { return 0 }
    return amount
    }
private var kiwiAmount: Int {
    guard let amount = getFruits[.kiwi] else { return 0 }
    return amount
    }
private var mangoAmount: Int {
    guard let amount = getFruits[.mango] else { return 0 }
    return amount
    }
```
**수정 후**
```swift
private var strawberryAmount: Int {
    return fruitStore.fruits[.strawberry] ?? 0
}
private var bananaAmount: Int {
    return fruitStore.fruits[.banana] ?? 0
}
private var pineappleAmount: Int {
    return fruitStore.fruits[.pineapple] ?? 0
}
private var kiwiAmount: Int {
    return fruitStore.fruits[.kiwi] ?? 0
}
private var mangoAmount: Int {
    return fruitStore.fruits[.mango] ?? 0
}
```
* 어제 말씀해주셨을 때 이 부분을 없애거나 줄일 수 있을것 같다고 얘기하셨는데 딕셔너리 옵셔널 값을 벗기는 작업이라 수정하는 과정에서 필수적으로 남겨야 된다고 생각합니다.
* 이 부분을 없애거나 수정하기 위해서는 FruitStore내의 프로퍼티나 메소드로 남겨서 JuiceMakerViewController내의 fruitStore로 접근하여 가져오는 것 이외에는 방법이 없을 것 같은데 그런 방법으로 JuiceMakerViewController의 코드 수를 줄이는 방법을 말씀하신걸까요?
* 저희가 사용하고 있는 MVC패턴의 큰 단점이 코드가 많아질수록 뷰컨트롤러의 역할과 코드수가 늘어난다는 것인데 이 부분때문에 위와 같은 프로퍼티들은 모델에 남기는게 좋은걸까요?

seu0313 commented on JuiceMaker/JuiceMaker/Model/JuiceProtocol.swift:
```diff
@@ -0,0 +1,6 @@
+//  JuiceMaker - JuiceProtocol.swift
+//  Created by Vetto, 레옹아범 on 2023.1.13.
+
+protocol UpdatableFruitStock: AnyObject {
+    func updateStockDelegate(changeStock: [Fruit: Int])
+}
```
디스코드에서 말씀드린 것처럼 Delegate 패턴을 적용한 프로토콜의 경우
일반적으로 타입의 이름으로 ~Delegate, ~Listener, ~Handler로 하는 편 입니다 :)

```swift
protocol ExampleDelegate {
    // ...
}

protocol ExampleListener {
    // ...
}
```

seu0313 reviewed (approved):
쥬스메이커 3주간 수고하셨습니다 vetto, 레옹아범 :)
다음 프로젝트도 화이팅입니다 🎉🎉🎉

아래는 추가로 주신 질문에 대한 답입니다.

**Q1. JuiceMakerViewController 딕셔너리 부분**

> 어제 말씀해주셨을 때 이 부분을 없애거나 줄일 수 있을것 같다고 얘기하셨는데 딕셔너리 옵셔널 값을 벗기는 작업이라 수정하는 과정에서 필수적으로 남겨야 된다고 생각합니다.
이 부분을 없애거나 수정하기 위해서는 FruitStore내의 프로퍼티나 메소드로 남겨서 JuiceMakerViewController내의 fruitStore로 접근하여 가져오는 것 이외에는 방법이 없을 것 같은데 그런 방법으로 JuiceMakerViewController의 코드 수를 줄이는 방법을 말씀하신걸까요?
저희가 사용하고 있는 MVC패턴의 큰 단점이 코드가 많아질수록 뷰컨트롤러의 역할과 코드수가 늘어난다는 것인데 이 부분때문에 위와 같은 프로퍼티들은 모델에 남기는게 좋은걸까요?

A1. 질문 주신대로 과일 재고에서 해당하는 과일에 대한 재고를 가져오는 로직의 경우 Model에서 제공하는 것이 좋다고 생각합니다.

추가적으로 모든 Label의 텍스트를 갱신하는 메서드와 각각 Label의 텍스트를 갱신하는 메서드가 분리되어 있기 때문에 발생한 고민이라고 생각됩니다. 다음과 같이 두 메서드를 합치면 해결할 수 있어보입니다. (updateLabel 메서드를 제거하고 Label 갱신을 setLabel에서 수행)

```swift
private func setLabel() {
    let mappedLabels: [UILabel: Fruit] = [
        strawberryLabel: .strawberry,
        bananaLabel: .banana,
        kiwiLabel: .kiwi,
        pineappleLabel: .pineapple,
        mangoLabel: .mango,
    ]

    mappedLabels.forEach { label, fruit in
        label.text = "\(fruitStore.fruits[fruit] ?? 0)"
    }
}
```

```swift
// 클로저 (기본 - 후행 클로저 과정)

// 1)
mappedLabels.forEach { (label: UILabel, fruit: Fruit) -> Void in
    label.text = "\(fruitStore.fruits[fruit] ?? 0)"
}

// 2)
mappedLabels.forEach { label, fruit in
    label.text = "\(fruitStore.fruits[fruit] ?? 0)"
}

// 3)
mappedLabels.forEach {
    $0.text = "\(fruitStore.fruits[$1] ?? 0)"
}

// 4)
mappedLabels.forEach { $0.text = "\(fruitStore.fruits[$1] ?? 0)" }
```
